Add Dockerfile to build Python 3.8, 3.9, and 3.10 on Alpine Linux

stefanv opened the pull request:
Dockerfile to build container required for https://github.com/matthew-brett/multibuild/pull/430

stefanv committed:
Add Dockerfile to build Python 3.8, 3.9, and 3.10 on Alpine Linux

mattip commented:
Is using a single RUN command instead of breaking each command into a separate RUN the recommended way to structure a Dockerfile? I would think separate commands give a more layered approach to running the docker build.

mattip commented on docker/Dockerfile-alpine:
```diff
@@ -0,0 +1,118 @@
+    && rm -f get-pip.py \
+    && rm -rf ~/.cache
+
+CMD ["/bin/bash"]
```
We need to have something like this to choose the proper python image (taken from the other docker files)
```
# Run Python selection on way into image
COPY choose_python.sh /usr/bin/
ENTRYPOINT ["/usr/bin/choose_python.sh"]
``

stefanv replied:
This is now fixed.

mattip commented on docker/Dockerfile-alpine:
```diff
@@ -0,0 +1,118 @@
+    && apk del --no-network .fetch-deps \
+    && apk del --no-network .build-deps
+
+ENV PIP_ROOT_URL="https://bootstrap.pypa.io"
```
Why do you prefer using get-pip over using ensurepip?

stefanv replied:
This is what is being used in the other images, and also in the official Python docker images.  But I have no preference, personally.

stefanv replied:
I've now switched over to ensurepip.

adeak replied:
README.md says
> Pip installed for each Python via get-pip.py.

Does this also need updating?

Update: probably not, as this is just for alpine.

stefanv commented:
> Is using a single RUN command instead of breaking each command into a separate RUN the recommended way to structure a Dockerfile? I would think separate commands give a more layered approach to running the docker build.

Yes, this is purposeful.  It turns out that adding layers is very costly in terms of image size, and the best way to keep them small is to combine commands.

stefanv committed:
Set choose_python as entrypoint

Also configure /usr/bin/python to link to a version of Python that can
run virtualenv, install virtualenv, and link it to where choose_python
expects it to be.

stefanv committed:
Switch from get-pip to ensurepip

matthew-brett commented:
Nice - thanks for doing this ...

matthew-brett commented:
Test failure looks unrelated?   Does this need a Travis-CI entry?

stefanv committed:
Add Travis matrix entry for Alpine

stefanv commented:
I added a matrix entry for Alpine.

mattip commented:
Cool, thanks. Is there a way to confirm that the container is usable in https://github.com/multi-build/multibuild/pull/430 (note that the repo has moved)

matthew-brett commented:
I guess the trick would be to put the image up somewhere, so we could use it to test against the PR ...

mattip commented:
Probably the easiest way to do that is to merge this, so in it goes.

mattip commented:
The docker image is up https://hub.docker.com/u/multibuild. It is called "focal_alpine", maybe it should be alpine_x86_64 like the focal_x86_64 one. Is there a chance we will see an alpine_arm64v8 or alpine_i686 ?
